server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java

```diff
@@ -481,29 +481,42 @@ private Builder initializeEmpty(IndexMetadata indexMetadata, UnassignedInfo unas
                         )
                     );
                 }
-                shards.put(shardNumber, indexShardRoutingBuilder.build());
+                shards[shardNumber] = indexShardRoutingBuilder;
             }
             return this;
         }
 
         public Builder addReplica() {
-            for (var shardNumber : shards.keySet()) {
-                ShardId shardId = new ShardId(index, shardNumber);
+            if (shards == null) {
```

Comment:
AFAICS, we could instead

`assert shards != null`?

Likewise for the null check in the loop?

Reply:
++ done